File: server/internal/domain/board/labels.go
package board

import (
	"slices"
	"strings"
)

// TeamLabelPrefix namespaces the labels the bundled directory owns.
const TeamLabelPrefix = "Team::"

var legacyStatusListKeys = map[string]string{
	"Wating":  "wating",
	"Waiting": "wating",
	"Inbox":   "inbox",
	"To Do":   "todo",
	"Todo":    "todo",
	"Doing":   "doing",
	"Review":  "review",
	"Closed":  "closed",
}

var DeprecatedLabels = []string{
	"Doing", "Inbox", "Review", "Status::Inbox", "To Do", "Todo", "Wating",
	"組別::總召", "組別::行政", "組別::開發",
}

// DeprecatedLabel reports whether a label belongs to the legacy workflow or to
// the lifecycle namespace the board manages itself. It is the single source of
// truth for both the read filter and the write guard.
func DeprecatedLabel(name string) bool {
	if _, legacyStatus := LegacyStatusListKey(name); legacyStatus {
		return true
	}
	if strings.HasPrefix(name, "Status::") {
		return true
	}
	return slices.Contains(DeprecatedLabels, name)
}

// ReservedLabel reports whether a label is owned by the board configuration
// rather than by users, and therefore cannot be created, renamed, or deleted.
//
// The whole Team:: prefix is reserved, not only the configured names: a
// user-created Team::新組 would not match issueTeam during sync, so it would
// behave like an ordinary label until someone added it to board-directory.yml,
// and would then start reassigning cards.
func ReservedLabel(name string, teamLabels []string) bool {
	trimmed := strings.TrimSpace(name)
	if trimmed == "" {
		return true
	}
	if strings.HasPrefix(trimmed, TeamLabelPrefix) {
		return true
	}
	return DeprecatedLabel(trimmed) || slices.Contains(teamLabels, trimmed)
}

func LegacyStatusListKey(label string) (string, bool) {
	key, ok := legacyStatusListKeys[label]
	return key, ok
}

func NormalizeLabels(labels []string) ([]string, bool) {
	result := make([]string, 0, len(labels))
	seen := make(map[string]struct{}, len(labels))
	for _, raw := range labels {
		label := strings.TrimSpace(raw)
		if label == "" {
			return nil, false
		}
		if _, exists := seen[label]; exists {
			continue
		}
		seen[label] = struct{}{}
		result = append(result, label)
	}
	return result, true
}

func CanonicalLabels(existing []string, teamLabel string, teamLabels []string) []string {
	reserved := make(map[string]struct{}, len(teamLabels)+len(DeprecatedLabels))
	for _, label := range teamLabels {
		reserved[label] = struct{}{}
	}
	for _, label := range DeprecatedLabels {
		reserved[label] = struct{}{}
	}
	labels := make([]string, 0, len(existing)+1)
	for _, label := range existing {
		_, isReserved := reserved[label]
		_, isLegacyStatus := LegacyStatusListKey(label)
		if !isReserved && !isLegacyStatus && !strings.HasPrefix(label, "Status::") {
			labels = append(labels, label)
		}
	}
	if teamLabel != "" {
		labels = append(labels, teamLabel)
	}
	return labels
}

File: web/src/features/board/labels.ts
import type { Bootstrap } from "./model";

/**
 * Mirror of server/internal/domain/board/labels.go. Keep both in sync;
 * labels.test.ts mirrors that file's table so drift shows up as a test diff.
 */
export const teamLabelPrefix = "Team::";

export const deprecatedLabels = new Set(["Wating", "Waiting", "Inbox", "To Do", "Todo", "Doing", "Review", "Closed", "組別::總召", "組別::行政", "組別::開發"]);

/** Legacy workflow labels and lifecycle labels the board manages itself. */
export function isDeprecatedLabel(name: string) {
	return deprecatedLabels.has(name) || name.startsWith("Status::");
}

/**
 * Labels owned by the board configuration rather than by users: the whole
 * `Team::` prefix plus every configured team label, and the legacy workflow
 * names. The prefix matters on its own — a user-created `Team::新組` would
 * behave like an ordinary label until someone adds it to board-directory.yml,
 * and would then start reassigning cards.
 */
export function isReservedLabel(bootstrap: Bootstrap, name: string) {
	const trimmed = name.trim();
	if (!trimmed) return true;
	if (trimmed.startsWith(teamLabelPrefix)) return true;
	if (isDeprecatedLabel(trimmed)) return true;
	return bootstrap.teams.some((team) => team.gitLabLabel === trimmed);
}

/** Card labels reduced to general labels plus exactly the active team's label. */
export function canonicalClientLabels(bootstrap: Bootstrap, labels: string[], teamKey: string) {
	const teamLabels = new Set(bootstrap.teams.map((team) => team.gitLabLabel));
	const general = labels.filter((label) => !teamLabels.has(label) && !isDeprecatedLabel(label));
	const teamLabel = bootstrap.teams.find((team) => team.key === teamKey && team.active)?.gitLabLabel;
	return [...general, ...(teamLabel ? [teamLabel] : [])];
}
